### PR TITLE
infoschema: fix bug of column size of set and enum type

### PR DESCRIPTION
**Pull request by imtbkcat**
## What have you changed? (mandatory)
I add code to fix bug when query column size of set and enum type.
Originally the column size of set and enum type while be default value 2^64 - 1. (See mysql/util:65)
This is different from MySQL's result.
For example, in MySQL 5.7.  

    mysql> create table bug22613 ( s set('a','bc','def','ghij') default NULL, t enum('a', 'ab', 'cdef'), s2 
    SET('1','2','3','4','1585','ONE','TWO','Y','N','THREE'));
    Query OK, 0 rows affected (0.11 sec)
    
    mysql> SELECT column_name,         character_maximum_length  FROM   
    information_schema.columns  WHERE  table_schema = Database()         AND         table_name = 
    'bug22613'         AND         column_name = 't';
    +-------------+--------------------------+
    | column_name | character_maximum_length |
    +-------------+--------------------------+
    | t           |                        4 |
    +-------------+--------------------------+
    1 row in set (0.00 sec)
But in TiDB:

    mysql> SELECT column_name,         character_maximum_length  FROM   
    information_schema.columns  WHERE  table_schema = Database()         AND         table_name = 
    'bug22613'         AND         column_name = 't';
    +-------------+--------------------------+
    | column_name | character_maximum_length |
    +-------------+--------------------------+
    | t           |     18446744073709551615 |
    +-------------+--------------------------+
    1 row in set (0.00 sec)

I fix this bug by adding code to calculate size of set and enum like MySQL.

## What is the type of the changes? (mandatory)

- Bug fix (non-breaking change which fixes an issue)

## How has this PR been tested? (mandatory)
unit test

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)
No

## Does this PR affect tidb-ansible update? (mandatory)
No

## Does this PR need to be added to the release notes? (mandatory)
No

## Refer to a related PR or issue link (optional)
Fix https://github.com/pingcap/tidb/issues/7297

### CODE REVIEW

File: infoschema/tables_test.go [hunk 1]
```diff
@@ -77,6 +77,17 @@ func (s *testSuite) TestDataForTableRowsCountField(c *C) {
...
+	//Test for length of enmu and set
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t ( s set('a','bc','def','ghij') default NULL, e1 enum('a', 'ab', 'cdef'), s2 SET('1','2','3','4','1585','ONE','TWO','Y','N','THREE'))")
+	tk.MustQuery("select column_name, character_maximum_length from information_schema.columns where table_schema=Database() and table_name = 't' and column_name = 's'").Check(
+		testkit.Rows("s 13"))
+	tk.MustQuery("select column_name, character_maximum_length from information_schema.columns where table_schema=Database() and table_name = 't' and column_name = 's2'").Check(
+		testkit.Rows("s2 30"))
+	tk.MustQuery("select column_name, character_maximum_length from information_schema.columns where table_schema=Database() and table_name = 't' and column_name = 'e1'").Check(
+		testkit.Rows("e1 4"))
+	//
```
Comment (shenli) [inline infoschema/tables_test.go]: Remove this line.

File: infoschema/tables.go [hunk 2]
```diff
@@ -797,6 +797,30 @@ func dataForColumnsInTable(schema *model.DBInfo, tbl *model.TableInfo) [][]types
...
+				sumLen += len(ele)
+			}
+			if len(col.Elems) == 0 {
+				cnt = 0
+			} else {
+				cnt = len(col.Elems) - 1
+			}
+			colLen = sumLen + cnt
+		}
+		if col.Tp == mysql.TypeEnum {
```
Comment (shenli) [inline infoschema/tables.go]: else if

File: infoschema/tables.go [hunk 3]
```diff
@@ -797,6 +797,30 @@ func dataForColumnsInTable(schema *model.DBInfo, tbl *model.TableInfo) [][]types
...
+			//exapmle: In MySQL enum('a', 'ab', 'cdef') has length 4, This is because
+			//the longest string in the enum is 'cdef'
```
Comment (jackysp) [inline infoschema/tables.go]: ditto

File: infoschema/tables.go [hunk 4]
```diff
@@ -797,6 +797,30 @@ func dataForColumnsInTable(schema *model.DBInfo, tbl *model.TableInfo) [][]types
 		if colLen == types.UnspecifiedLength {
 			colLen = defaultFlen
 		}
+		if col.Tp == mysql.TypeSet {
+			//example: In MySQL set('a','bc','def','ghij') has length 13, This is because
+			//len('a')+len('bc')+len('def')+len('ghij')+len(ThreeComma)=13
```
[the remark this answers is not in the document]
Comment (zhexuany) [reply infoschema/tables.go]: It is better to add a reference link.

[the remark this answers is not in the document]
Comment (imtbkcat) [reply infoschema/tables.go]: Ok, I'm looking for MySQL's method.

**Comment by imtbkcat** [comment]
PTAL @jackysp @zhexuany @shenli

File: infoschema/tables.go [hunk 5]
```diff
@@ -797,6 +797,31 @@ func dataForColumnsInTable(schema *model.DBInfo, tbl *model.TableInfo) [][]types
 		if colLen == types.UnspecifiedLength {
 			colLen = defaultFlen
 		}
+		if col.Tp == mysql.TypeSet {
+			// Example: In MySQL set('a','bc','def','ghij') has length 13, because
+			// len('a')+len('bc')+len('def')+len('ghij')+len(ThreeComma)=13
+			// Reference link: https://bugs.mysql.com/bug.php?id=22613
+			sumLen, cnt := 0, 0
```
Comment (crazycs520) [inline infoschema/tables.go]: how about this:
```go
colLen = 0
for _, ele := range col.Elems {
	colLen += len(ele)
}
if len(col.Elems) != 0 {
	colLen = len(col.Elems) - 1
}
```

Comment (imtbkcat) [reply infoschema/tables.go]: @crazycs520 
it looks grate. But it should be:  

    if len(col.Elems) != 0 {
	    colLen += (len(col.Elems) - 1)
    }

Comment (crazycs520) [reply infoschema/tables.go]: ye~😢

File: infoschema/tables_test.go [hunk 6]
```diff
@@ -77,6 +77,16 @@ func (s *testSuite) TestDataForTableRowsCountField(c *C) {
 	tk.MustExec("create user xxx")
 	tk.MustExec("flush privileges")
 
+	//Test for length of enum and set
```
Comment (winoros) [inline infoschema/tables_test.go]: Add space. `// Test....`

**Comment by imtbkcat** [comment]
PTAL @winoros @crazycs520

**Pull request merged**